_posts/2020-03-16-dask-usage-at-blue-yonder.markdown

```diff
@@ -1,131 +1,131 @@
+  Our algorithms rely on source datasets that are partitioned in a particular way. It is not always
+  the case that we have the data available in a suitable partitioning, for instance, if the data results
+  from a computation that used a different partitioning. In this case, we have to repartition the data.
+  This works by reading the dataset as a Dask distributed dataframe (ddf), repartitioning the ddf using
+  network shuffle, and writing it out again to a dataset.
+
+  ## Dask cluster setup at Blue Yonder
+
+  We run a somewhat unique setup of Dask clusters that is partly driven by the specific requirements
+  of our domain and partly historically grown. We are running a large number of individual clusters
+  that are homogeneous in size with many of the clusters dynamically scaled. The peculiarities of this
+  setup have in some cases triggered edge cases and uncovered bugs, which lead us to submit a number
+  of upstream pull requests.
+
+  ### Cluster separation
+
+  We are running separate Dask clusters for separate customers in order to isolate the customers from each other.
+  For the same reason, we also run separate clusters for production, staging, and development environments. 
+  But it does not stop there. For performance reasons, we install the Python packages holding the code needed
+  for the computations on each worker. Since the different components of our compute pipeline are implemented
+  and maintained as completely separate software packages whose requirements are not synchronized, this also
+  means we have to run a separate Dask cluster for each of the components. This leads to us operating more than
+  ten Dask clusters per customer and environment, with most of the time, only one of the clusters being active
+  and computing something. While this leads to overhead in terms of administration and hardware resouces,
+  it also gives us a lot of flexibility. For instance, we can update the software on the cluster of one part of the compute pipeline
+  while another part of the pipeline is computing something on a different cluster.
+
+  ### Some numbers
+
+  The number and size of the workers varies from cluster to cluster depending on the degree of parallelism
+  of the computation being performed, its resource requirements, and the available timeframe for the computation.
+  At the time of writing this, we are running more than 500 distinct clusters.
+  They have between one and 225 workers, with worker size varying between 1GiB and 64GiB of memory.
+
+  ## Cluster scaling and resilience
+
+  To improve manageability, resilience, and resource utilization, we run the Dask clusters on top
+  of [Apache Mesos](http://mesos.apache.org/)/[Aurora](http://aurora.apache.org/). This
+  means it is really easy to commission or decommission Dask clusters and to add or remove workers to/from
+  an existing Dask cluster. At the moment, we are starting a migration from Mesos/Aurora towards 
+  [Kubernetes](https://kubernetes.io/)
```

Comment:
given that the article focus is on "usage of desk at Blue Yonder", my opinion would be to remove the last line.

Reply:
Yes, it's pretty unrelated. I somehow want to mention Kubernetes, as it's the Future. I'll just add it to the list of cluster managers in the sentence before.